From 10260a92e1948880440d965bda984be4f17d7371 Mon Sep 17 00:00:00 2001
From: Hsieh Hung-En <hungen3108@gmail.com>
Date: Sat, 31 Jan 2026 00:00:13 +0800
Subject: [PATCH 004/385] UPSTREAM: ASoC: es8328: Fix DAC deemphasis control
 handling

The DAC deemphasis control updated the hardware before updating the cached
state, causing the previous setting to be applied.

Update the cached deemphasis state first and then apply the setting.

Also check and propagate errors from es8328_set_deemph() in hw_params().

Signed-off-by: Hsieh Hung-En <hungen3108@gmail.com>
Link: https://patch.msgid.link/20260130160017.2630-2-hungen3108@gmail.com
Signed-off-by: Mark Brown <broonie@kernel.org>

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 sound/soc/codecs/es8328.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/sound/soc/codecs/es8328.c b/sound/soc/codecs/es8328.c
index 38340f2922826..46868b7924a0b 100644
--- a/sound/soc/codecs/es8328.c
+++ b/sound/soc/codecs/es8328.c
@@ -163,12 +163,11 @@ static int es8328_put_deemph(struct snd_kcontrol *kcontrol,
 	if (es8328->deemph == deemph)
 		return 0;
 
+	es8328->deemph = deemph;
 	ret = es8328_set_deemph(component);
 	if (ret < 0)
 		return ret;
 
-	es8328->deemph = deemph;
-
 	return 1;
 }
 
@@ -530,7 +529,9 @@ static int es8328_hw_params(struct snd_pcm_substream *substream,
 			return ret;
 
 		es8328->playback_fs = params_rate(params);
-		es8328_set_deemph(component);
+		ret = es8328_set_deemph(component);
+		if (ret < 0)
+			return ret;
 	} else {
 		ret = snd_soc_component_update_bits(component, ES8328_ADCCONTROL4,
 						    ES8328_ADCCONTROL4_ADCWL_MASK,
-- 
2.52.0

